File: Python/Binary Tree Postorder Traversal.py
# Given a binary tree, return the postorder traversal of its nodes' values.

# For example:
# Given binary tree {1,#,2,3},
   # 1
    # \
     # 2
    # /
   # 3
# return [3,2,1].

# Note: Recursive solution is trivial, could you do it iteratively?

# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# method 1
class Solution:
    # @param root, a tree node
    # @return a list of integers
    def postorderTraversal(self, root):
		s = []
		result = []
		rt = []
		if root == None:
			return result
		s.append(root)
		while len(s) != 0:
			current = s.pop()
			result.append(current)
			if current.left != None:
				s.append(current.left)
			if current.right != None:
				s.append(current.right)
		while len(result) != 0:
			rt.append(result.pop().val)
		return rt
		
#method 2

File: Java/Binary_Tree_Postorder_Traversal.java
class Binary_Tree_Postorder_Traversal{
	public List<Integer> postorderTraversal(TreeNode root) {

	}
	private void recursive(ArrayList<Integer> res, TreeNode root){
		if(root == null)
			return;
		if(root.left != null)
			recursive(root.left);
		if(root.right != null)
			recursive(root.right);
		res.add(root.val);
		return;
	}
	private void iterative(ArrayList<Integer> res, TreeNode root){
		if(root == null)
			return;
		LinkedList<TreeNode> stack = new LinkedList<Integer>();
		TreeNode lastVisited = null;
		while(!stack.isEmpty()||root != null){
			if(root != null){
				stack.push(root);
				root = root.left;
			}
			else{
				TreeNode temp = stack.peek();
				if(temp.right != null && temp.right != lastVisited){
					root = temp.right;
				}
				else{
					res.add(temp.val);
					stack.pop();
					lastVisited = temp;
				}
			}
		}
	}
}